File: scala/hackathon/palindrome.scala
def palindrome(n: Int) = { n.toString() == n.toString().reverse }

def palindromePairs() = for (x <- (1 to 999); y <- (1 to 999) if palindrome(x * y)) yield (x, y)

def maxPair(p1: (Int, Int), p2: (Int, Int)) = { if ((p1._1 + p1._2) > (p2._1 + p2._2)) p1 else p2 }

def main() = {
    val p = palindromePairs().foldLeft((0,0))(maxPair)
    println("largest pair: " + p)
    println("palindrome: " + p._1 * p._2)
}

File: haskell/hackathon/palindrome.hs
palindrome n = s == reverse s
    where s = (show n)

nums = [100..999]

palindromePairs = [(x,y) | x <- nums, y <- nums, palindrome (x * y)]

maxPair (x1,y1) (x2,y2)
    | x1 + y1 > x2 + y2 = (x1,y1)
    | otherwise = (x2,y2)
    
largest = foldl maxPair (0,0) palindromePairs

main = do
    let x = largest
    putStr "largest is "
    print x
    let y = (fst x) * (snd x)
    putStr "palindrome: "
    print y